shcomplete/model2predict.py

```diff
@@ -0,0 +1,208 @@
+            self.command2index[self.vocab[i]] = i
+
+    @property
+    def size(self):
+        """
+        Return the number of prefixes in the vocabulary.
+        """
+        return len(self.vocab)
+
+    def trie(self, path_to_vocab):
+        """
+        Return the vocabulary stored in a trie structure.
+        """
+        with open(path_to_vocab, "r") as f:
+            content = f.read().split("\n")
+        return build_trie(content)
+
+    def encode(self, batch_of_data, vocab_trie, seq_len, target=False):
+        """
+        Encode sequences of commands into numpy arrays.
+        """
+        if target:
+            y = np.zeros(len(batch_of_data))
+            for i, command in enumerate(batch_of_data):
+                try:
+                    next_prefix = vocab_trie.longest_prefix(command)[0]
+                    y[i] = self.command2index[next_prefix]
+                except KeyError:
+                    y[i] = self.command2index["UNK"]
+            return y
+        else:
+            X = np.zeros((len(batch_of_data), seq_len))
+            for i, sequence in enumerate(batch_of_data):
+                for j, cmd in enumerate(sequence):
+                    try:
+                        prefix = vocab_trie.longest_prefix(cmd)[0]
+                        X[i, j] = self.command2index[prefix]
+                    except KeyError:
+                        X[i, j] = self.command2index["UNK"]
+            return X
+
+    def decode(self, X, reduction=False):
+        """
+        Decode the numpy array X and return the corresponding command.
+        """
+        if reduction:
+            X = X.argmax(axis=-1)
+        sequence = "\n".join(self.index2command[i] for i in X)
+        return sequence
+
+
+def generator(path_to_corpus, path_to_vocab, batch_size, seq_len):
+    """
+    Return a random batch of data to feed fit_generator.
+    The ouput is a tuple (sequences, outputs)
+    """
+    vocab = Vocabulary(path_to_vocab)
+    vocab_trie = vocab.trie(path_to_vocab)
+    with open(path_to_corpus) as f:
+        histories = f.read().split("\n\n")
+        while True:
+            sequences = []
+            next_commands = []
+            while len(sequences) < batch_size:
+                id_hist = randint(len(histories)-1)
+                history = histories[id_hist].split("\n")
+                id_cmd = randint(len(history[:-seq_len]))
+                target = vocab_trie.longest_prefix(history[id_cmd + seq_len])[0]
+                if target:
+                    sequence = [cmd.rstrip() for cmd in history[id_cmd:id_cmd + seq_len]]
+                    sequences.append(sequence)
+                    next_commands.append(history[id_cmd + seq_len])
+            X = vocab.encode(sequences, vocab_trie, seq_len)
+            y = vocab.encode(next_commands, vocab_trie, seq_len, target=True)
+            yield X, y[:, np.newaxis, np.newaxis]
+
+
+def generate_model(args):
+    """
```

Comment:
I see exactly the same NN code here as in misprints. You need to refactor this to be DRY